academy/3-my-own-chain/gas-meter.md

```diff
@@ -54,44 +54,56 @@ Before diving into the specifics, ask yourself:
+
+As a checkers blockchain creator, your goal may be to have as many on-going games as possible. Adding costs sounds counter to this goal. However here the goal is to optimize potential congestion at the margin. If there is little activity, then the gas price will go down, and these additional costs will be trivial for players anyway. Conversely, if there is a lot of network activity, the gas price will go up, and whether you have put additional costs or not, players will still be less likely to participate.
+
 ## Add handling
 
-Add a line that consumes the designated amount of gas in each relevant handler:
+Add a line that consumes or refunds the designated amount of gas in each relevant handler:
 
 1. When handling a game creation:
 
-    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/63370efe/x/checkers/keeper/msg_server_create_game.go#L45]
+    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/gas-meter/x/checkers/keeper/msg_server_create_game.go#L46]
     ctx.GasMeter().ConsumeGas(types.CreateGameGas, "Create game")
     ```
 
 2. When handling a move:
 
-    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/63370efe/x/checkers/keeper/msg_server_play_move.go#L94]
+    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/gas-meter/x/checkers/keeper/msg_server_play_move.go#L88]
     ctx.GasMeter().ConsumeGas(types.PlayMoveGas, "Play a move")
     ```
 
-3. When handling a game rejection:
+3. When handling a game rejection, you make sure that you are not refunding more than what has already been consumed:
 
-    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/63370efe/x/checkers/keeper/msg_server_reject_game.go#L52]
-    ctx.GasMeter().ConsumeGas(types.RejectGameGas, "Reject game")
+    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/gas-meter/x/checkers/keeper/msg_server_reject_game.go#L46-L50]
+    refund := uint64(types.RejectGameRefundGas)
+    if consumed := ctx.GasMeter().GasConsumed(); consumed < refund {
+        refund = consumed
+    }
+    ctx.GasMeter().RefundGas(refund, "Reject game")
     ```
 
 You do not meter gas in your `EndBlock` handler because it is **not** called by a player sending a transaction. Instead, it is a service rendered by the network. If you want to account for the gas cost of a game expiration, you have to devise a way to pre-collect it from players as part of the other messages.
 
 <HighlightBox type="tip">
 
-As part of your code optimization, avoid calling `ConsumeGas` with a fixed gas cost (for instance `k`) from within a loop. Each pass of the loop uses computation resources (`c`) on each node. If you know the number of times your code loops (`n`), you know running the full loop will use `n*c` computation resources.
-<br></br>
```

Comment:
Just to note, I've been manually adding a lot of these breaks (line 94) in highlight boxes to add a space between regular paragraph formatting, which doesn't show up in them. I don't know if this is one that I added, but for consistency with the rest it should probably be reintroduced.

Reply:
Yes, I had to go through a ton of merge conflicts and this mistake probably slipped through...